service_capacity_modeling/models/org/netflix/elasticsearch.py

```diff
@@ -0,0 +1,471 @@
+
+    needed_disk = math.ceil(
+        (1.0 / desires.data_shape.estimated_compression_ratio.mid)
+        * desires.data_shape.estimated_state_size_gib.mid
+        * copies_per_region,
+    )
+
+    # Rough estimate of how many instances we would need just for the the CPU
+    # Note that this is a lower bound, we might end up with more.
+    needed_cores = math.ceil(
+        max(1, needed_cores // (instance.cpu_ghz / desires.core_reference_ghz))
+    )
+    rough_count = math.ceil(needed_cores / instance.cpu)
+
+    # Generally speaking we want fewer than some number of reads per second
+    # hitting disk per instance. If we don't have many reads we don't need to
+    # hold much data in memory.
+    instance_rps = max(1, reads_per_second // rough_count)
+    disk_rps = instance_rps * _es_io_per_read(max(1, needed_disk // rough_count))
+    rps_working_set = min(1.0, disk_rps / max_rps_to_disk)
+
+    # If disk RPS will be smaller than our target because there are no
+    # reads, we don't need to hold as much data in memory
+    needed_memory = min(working_set, rps_working_set) * needed_disk
+
+    # Now convert to per zone
+    needed_cores = needed_cores // zones_per_region
+    needed_disk = needed_disk // zones_per_region
+    needed_memory = int(needed_memory // zones_per_region)
+    logger.debug(
+        "Need (cpu, mem, disk, working) = (%s, %s, %s, %f)",
+        needed_cores,
+        needed_memory,
+        needed_disk,
+        working_set,
+    )
+
+    return CapacityRequirement(
+        requirement_type="elasticsearch-data-zonal",
+        core_reference_ghz=desires.core_reference_ghz,
+        cpu_cores=certain_int(needed_cores),
+        mem_gib=certain_float(needed_memory),
+        disk_gib=certain_float(needed_disk),
+        network_mbps=certain_float(needed_network_mbps),
+        context={
+            "working_set": min(working_set, rps_working_set),
+            "rps_working_set": rps_working_set,
+            "disk_slo_working_set": working_set,
+            "replication_factor": copies_per_region,
+            "compression_ratio": round(
+                1.0 / desires.data_shape.estimated_compression_ratio.mid, 2
+            ),
+            "read_per_second": reads_per_second,
+        },
+    )
+
+
+def _upsert_params(cluster, params):
+    if cluster.cluster_params:
+        cluster.cluster_params.update(params)
+    else:
+        cluster.cluster_params = params
+
+
+# pylint: disable=too-many-locals
+def _estimate_elasticsearch_cluster_zonal(
+    instance: Instance,
+    drive: Drive,
+    desires: CapacityDesires,
+    zones_per_region: int = 3,
+    copies_per_region: int = 3,
+    max_local_disk_gib: int = 4096,
+    max_regional_size: int = 240,
+    max_rps_to_disk: int = 500,
+) -> Optional[CapacityPlan]:
+
+    # Netflix Elasticsearch doesn't like to deploy on really small instances
+    if instance.cpu < 2 or instance.ram_gib < 14:
+        return None
+
+    # (FIXME): Need elasticsearch input
```

Comment:
Instances with attached SSDs worked for us so far. No plans to support EBS.

Reply:
This is true for data nodes, master nodes run on EBS-based instances like r5.

Reply:
yeah right now the model is only spitting out the data clusters, can add masters in and routing in a followup.